### PR TITLE
Swagger2 Extensions

### PR DESCRIPTION
**Pull request by FrancisLennon17**
Added extensions, marshalling/unmarshalling and conversion to and from V3 for:
- Root swagger
- Path 
- Operation
- Parameter (& v3 Requestbody)
- Responses
- SecurityDefinitions

### CODE REVIEW

File: openapi2/openapi2.go [hunk 1]
```diff
@@ -179,14 +291,58 @@ type Header struct {
...
+	StripNonCustomExtensions(&s.Extensions)
+	*securityScheme = SecurityScheme(s)
+	return nil
+}
+
+func (s SecurityScheme) MarshalJSON() ([]byte, error) {
+	return marshallFromMap(s.Extensions, reflect.ValueOf(s), reflect.TypeOf(s))
+}
+
+func StripNonCustomExtensions(extensions *map[string]interface{}) {
```
Comment (fenollp) [inline openapi2/openapi2.go:325]: I don't think you need to use a pointer to a map as map are already references, no?

File: openapi2/openapi2.go [hunk 2]
```diff
@@ -43,16 +65,49 @@ func (swagger *Swagger) AddOperation(path string, method string, operation *Oper
...
+	if err := json.Unmarshal([]byte(b), &p.Extensions); err != nil {
+		return err
+	}
+	StripNonCustomExtensions(&p.Extensions)
+	*pathItem = PathItem(p)
+	return nil
+}
+
+func (p PathItem) MarshalJSON() ([]byte, error) {
+	return marshallFromMap(p.Extensions, reflect.ValueOf(p), reflect.TypeOf(p))
```
Comment (fenollp) [inline openapi2/openapi2.go:110]: Why not use the jsoninfo package that's widely used in openapi3 package?

Comment (FrancisLennon17) [reply openapi2/openapi2.go:110]: Honestly I was a little intimidated by it at first, but I think I've a better grasp of it now. I'll make that change today 👍

**Comment by FrancisLennon17** [comment]
@fenollp Could you take another look. jsoninfo package made that a lot easier 🤦

File: openapi2/openapi2.go [hunk 3]
```diff
@@ -11,10 +11,13 @@ import (
 	"fmt"
 	"net/http"
 
+	"github.com/getkin/kin-openapi/jsoninfo"
+
```
Comment (fenollp) [inline openapi2/openapi2.go:15]: Can you remove the newline here?

File: openapi2conv/openapi2_conv_test.go [hunk 4]
```diff
@@ -22,7 +22,7 @@ func TestConvOpenAPIV3ToV2(t *testing.T) {
 }
 
 func TestConvOpenAPIV2ToV3(t *testing.T) {
-	var swagger2 openapi2.Swagger
+	swagger2 := openapi2.Swagger{}
```
Comment (fenollp) [inline openapi2conv/openapi2_conv_test.go:25]: Could you revert this? I'm not seeing what it enables.

File: openapi2/openapi2.go [hunk 5]
```diff
@@ -43,7 +54,22 @@ func (swagger *Swagger) AddOperation(path string, method string, operation *Oper
 	pathItem.SetOperation(method, operation)
 }
 
+func (swagger *Swagger) AddPathExtensions(path string, extensionProps openapi3.ExtensionProps) {
```
Comment (fenollp) [inline openapi2/openapi2.go:57]: It looks like this is only called from the conversion package and could very well be implemented there.
How do you see this function being useful to other callers? Can we make this a private func within `openapi2_conv`?

**Comment by FrancisLennon17** [comment]
@fenollp Made the changes you requested there.

**Comment by FrancisLennon17** [comment]
Thank you @fenollp 🙏

**Comment by fenollp** [comment]
Thank you :)

**Pull request merged**